### DISCUSSION TITLE
Subscribe on changes.

### DISCUSSION THREAD

**Discussion by nosferatu500** [pmndrs/valtio#530] [Q&A] [answered]
I found it very useful to use `useSnapshot(store)` to re-render a component.

But when I using it like this

```
useSnapshot(store)
useSnapshot(store2)

render (
// other things
)
```

and modify 2 different stores at the same time, the above component will re-render 2 times. Are there any options to have the re-render fire once if both stores have changed?

**Comment by dai-shi** [comment, marked as answer]
How about this?
```js
const { store1: snap1, store2: snap2 } = useSnapshot(useMemo(() => proxy({ store1, store2 }), []);
```

If possible, I'd define `proxy({ store1, store2 })` outside component.

**Reply by nosferatu500** [reply]
Hmm...

Let me explain a little. I have a component with a data generator. I need to re-render this component every time any of my stores changes.

For example:
```
// Component
useSnapshot(store1)
useSnapshot(store2)
useSnapshot(store3)
useSnapshot(store4)
useSnapshot(store5)
useSnapshot(store6)


return (
   <div>
   myGenerator()
   </div>
)
```

// another.ts file
```
const myGenerator = () => {
   store1.data....
   store3.data....
   
   // some magic
   
   return newData
}
```

Is it better to use separate `useSnapshots` in this case, or is it better to combine them with the combination `useSnapshot(useMemo)`?

It's question just about BestPractice/RightWay things.

**Reply by dai-shi** [reply]
Then, create a combined store and make subscription (useSnapshot) scoped.

```js
const state = proxy({...})

  // in a component
  useSnapshot(state)

  // in another component
  useSnapshot(state.foo)
```

So, the recommendation is to structure your state so that you can just have one `useSnapshot` in a component. You can use sub-stores just like separate store.

**Reply by nosferatu500** [reply]
Yes, I just tried your example with the combination `useSnapshot(useMemo(), [])` and it works (re-render time) much faster than when I used multiple `useSnapshot`. 

>Then, create a combined store and make subscription (useSnapshot) scoped.

I didn't know that I can do this! (maybe I just didn't found it in the documentation). Thank you for the explanations!

**Reply by dai-shi** [reply]
Well, `useSnapshot(useMemo())` solution is something I made up just this time.
Again, the general recommendation is not to do it, but to create your store combined in the first place.

**Reply by nosferatu500** [reply]
I'm not sure, but maybe add a new rule to `eslint-plugin-valtio` to warn users about using multiple useSnapshots in a component?

**Reply by dai-shi** [reply]
Using multiple `useSnapshot`s should not be that a bad thing, though.
